# Lab For Experimentation

In [1]:
import warnings
from typing import Any, Literal

import numpy as np
import pandas as pd
import polars as pl
from rich.console import Console
from rich.theme import Theme

custom_theme = Theme(
    {
        "white": "#FFFFFF",  # Bright white
        "info": "#00FF00",  # Bright green
        "warning": "#FFD700",  # Bright gold
        "error": "#FF1493",  # Deep pink
        "success": "#00FFFF",  # Cyan
        "highlight": "#FF4500",  # Orange-red
    }
)
console = Console(theme=custom_theme)

# Visualization
# import matplotlib.pyplot as plt

# NumPy settings
np.set_printoptions(precision=4)

# Pandas settings
pd.options.display.max_rows = 1_000
pd.options.display.max_columns = 1_000
pd.options.display.max_colwidth = 600

# Polars settings
pl.Config.set_fmt_str_lengths(1_000)
pl.Config.set_tbl_cols(n=1_000)
pl.Config.set_tbl_rows(n=200)

warnings.filterwarnings("ignore")

# Black code formatter (Optional)
%load_ext lab_black

# auto reload imports
%load_ext autoreload
%autoreload 2

In [2]:
def go_up_from_current_directory(*, go_up: int = 1) -> None:
    """This is used to up a number of directories.

    Params:
    -------
    go_up: int, default=1
        This indicates the number of times to go back up from the current directory.

    Returns:
    --------
    None
    """
    import os
    import sys

    CONST: str = "../"
    NUM: str = CONST * go_up

    # Goto the previous directory
    prev_directory = os.path.join(os.path.dirname(__name__), NUM)
    # Get the 'absolute path' of the previous directory
    abs_path_prev_directory = os.path.abspath(prev_directory)

    # Add the path to the System paths
    sys.path.insert(0, abs_path_prev_directory)
    print(abs_path_prev_directory)


# Demo (Prevents ruff from removing the unused module import)
name: Any
category: Literal["A", "B", "C"]

In [3]:
go_up_from_current_directory(go_up=1)

/Users/mac/Desktop/Projects/Bike-Rental-Prediction


# Table of Contents
- [Dataset Preparation](#Dataset-Preparation)
  - [Load Data](#load-data)
  - [Validate Data](#validate-data)
- [Baseline Model](#Baseline-Model)
- [Baseline Model With Lagged Target Features](#baseline-model-with-lagged-target-features)
- [Gradient Boosting With Hyperparameters Tuning](#Gradient-Boosting-with-Hyperparameters-Tuning)
- [Conclusions](#Conclusions)

In [4]:
from src.utilities.data_validator import data_validator

# Dataset Preparation

### Load data

In [5]:
fp: str = "../../../../Documents/data_dump/bike_data/database.parquet"
data: pl.DataFrame = pl.read_parquet(fp)
console.print(f"Shape: {data.shape}", style="info")

data.head()

Shape: (13903, 16)

shape: (5, 16)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┬─────┐
│ dat ┆ sea ┆ yr  ┆ mnt ┆ hr  ┆ hol ┆ wee ┆ wor ┆ wea ┆ tem ┆ ate ┆ hum ┆ win ┆ cas ┆ regist ┆ cnt │
│ eti ┆ son ┆ --- ┆ h   ┆ --- ┆ ida ┆ kda ┆ kin ┆ the ┆ p   ┆ mp  ┆ --- ┆ dsp ┆ ual ┆ ered   ┆ --- │
│ me  ┆ --- ┆ i64 ┆ --- ┆ i64 ┆ y   ┆ y   ┆ gda ┆ rsi ┆ --- ┆ --- ┆ f64 ┆ eed ┆ --- ┆ ---    ┆ i64 │
│ --- ┆ i64 ┆     ┆ i64 ┆     ┆ --- ┆ --- ┆ y   ┆ t   ┆ f64 ┆ f64 ┆     ┆ --- ┆ i64 ┆ i64    ┆     │
│ str ┆     ┆     ┆     ┆     ┆ i64 ┆ i64 ┆ --- ┆ --- ┆     ┆     ┆     ┆ f64 ┆     ┆        ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ i64 ┆ i64 ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╪═════╡
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 0   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 3   ┆ 13     ┆ 16  │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 879 ┆ 1   ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 1   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 8   ┆ 32     ┆ 40  │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 2   ┆ 727 ┆     ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 01: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 2   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 5   ┆ 27     ┆ 32  │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 2   ┆ 727 ┆     ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 02: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 3   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.7 ┆ 0.0 ┆ 3   ┆ 10     ┆ 13  │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 879 ┆ 5   ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 03: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 4   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.7 ┆ 0.0 ┆ 0   ┆ 1      ┆ 1   │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 879 ┆ 5   ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆   

### Validate Data

In [6]:
# Validate the data
data_report: dict[str, Any] = data_validator(data)
console.print(data_report)

{
    'data_schema': {
        'numeric': {
            'season': 'Int64',
            'yr': 'Int64',
            'mnth': 'Int64',
            'hr': 'Int64',
            'holiday': 'Int64',
            'weekday': 'Int64',
            'workingday': 'Int64',
            'weathersit': 'Int64',
            'temp': 'Float64',
            'atemp': 'Float64',
            'hum': 'Float64',
            'windspeed': 'Float64',
            'casual': 'Int64',
            'registered': 'Int64',
            'cnt': 'Int64'
        },
        'string': {'datetime': 'String'}
    },
    'data_shape': {
        'total_rows': 13903,
        'total_columns': 16,
        'number_of_numeric_columns': 15,
        'number_of_string_columns': 1
    },
    'summary_statistics': {
        'numeric': [
            {
                'column': 'season',
                'mean': 2.26,
                'median': 2.0,
                'mode': [2],
                'std': 1.04,
                'variance': 1.07,
                'range': 3,
                'min': 1,
                'max': 4,
                'count': 13903,
                'missing_values': 0,
                'missing_pct': 0.0,
                'unique_values': 4
            },
            {
                'column': 'yr',
                'mean': 0.38,
                'median': 0.0,
                'mode': [0],
                'std': 0.48,
                'variance': 0.24,
                'range': 1,
                'min': 0,
                'max': 1,
                'count': 13903,
                'missing_values': 0,
                'missing_pct': 0.0,
                'unique_values': 2
            },
            {
                'column': 'mnth',
                'mean': 5.65,
                'median': 5.0,
                'mode': [7, 5],
                'std': 3.22,
                'variance': 10.36,
                'range': 11,
                'min': 1,
                'max': 12,
                'count': 13903,
                'missing_values': 0,
                'missing_pct': 0.0,
                'unique_values': 12
            },
            {
                'column': 'hr',
                'mean': 11.55,
                'median': 12.0,
                'mode': [16, 17],
                'std': 6.91,
                'variance': 47.78,
                'range': 23,
                'min': 0,
                'max': 23,
                'count': 13903,
                'missing_values': 0,
                'missing_pct': 0.0,
                'unique_values': 24
            },
            {
                'column': 'holiday',
                'mean': 0.03,
                'median': 0.0,
                'mode': [0],
                'std': 0.16,
                'variance': 0.03,
                'range': 1,
                'min': 0,
                'max': 1,
                'count': 13903,
                'missing_values': 0,
                'missing_pct': 0.0,
                'unique_values': 2
            },
            {
                'column': 'weekday',
                'mean': 3.0,
                'median': 3.0,
                'mode': [6],
                'std': 2.01,
                'variance': 4.02,
                'range': 6,
                'min': 0,
                'max': 6,
                'count': 13903,
                'missing_values': 0,
                'missing_pct': 0.0,
                'unique_values': 7
            },
            {
                'column': 'workingday',
                'mean': 0.68,
                'median': 1.0,
                'mode': [1],
                'std': 0.46,
                'variance': 0.22,
                'range': 1,
                'min': 0,
                'max': 1,
                'count': 13903,
                'missing_values': 0,
                'missing_pct': 0.0,
                'unique_values': 2
            },
            {
                'column': 'weathersit',
                'mean': 1.42,
                'median': 1.0,
                'mod

In [7]:
data.head()

shape: (5, 16)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┬─────┐
│ dat ┆ sea ┆ yr  ┆ mnt ┆ hr  ┆ hol ┆ wee ┆ wor ┆ wea ┆ tem ┆ ate ┆ hum ┆ win ┆ cas ┆ regist ┆ cnt │
│ eti ┆ son ┆ --- ┆ h   ┆ --- ┆ ida ┆ kda ┆ kin ┆ the ┆ p   ┆ mp  ┆ --- ┆ dsp ┆ ual ┆ ered   ┆ --- │
│ me  ┆ --- ┆ i64 ┆ --- ┆ i64 ┆ y   ┆ y   ┆ gda ┆ rsi ┆ --- ┆ --- ┆ f64 ┆ eed ┆ --- ┆ ---    ┆ i64 │
│ --- ┆ i64 ┆     ┆ i64 ┆     ┆ --- ┆ --- ┆ y   ┆ t   ┆ f64 ┆ f64 ┆     ┆ --- ┆ i64 ┆ i64    ┆     │
│ str ┆     ┆     ┆     ┆     ┆ i64 ┆ i64 ┆ --- ┆ --- ┆     ┆     ┆     ┆ f64 ┆     ┆        ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ i64 ┆ i64 ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╪═════╡
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 0   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 3   ┆ 13     ┆ 16  │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 879 ┆ 1   ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 1   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 8   ┆ 32     ┆ 40  │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 2   ┆ 727 ┆     ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 01: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 2   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 5   ┆ 27     ┆ 32  │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 2   ┆ 727 ┆     ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 02: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 3   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.7 ┆ 0.0 ┆ 3   ┆ 10     ┆ 13  │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 879 ┆ 5   ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 03: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00: ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 00  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 201 ┆ 1   ┆ 0   ┆ 1   ┆ 4   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.2 ┆ 0.7 ┆ 0.0 ┆ 0   ┆ 1      ┆ 1   │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 879 ┆ 5   ┆     ┆     ┆        ┆     │
│ 1-0 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        ┆     │
│ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆   

### Comment

- Drop the columns with high cardinality
    - `datetime`

- Drop irrelevant columns
    - `casual` (used to create the target)
    - `registered` (used to create the target)
    - `atemp` (highly correlated with `temp`) 
    - `yr` (not relevant)


In [8]:
year: dict[str, Any] = data_report.get("summary_statistics").get("numeric")[1]
atemp: dict[str, Any] = data_report.get("summary_statistics").get("numeric")[9]
casual: dict[str, Any] = data_report.get("summary_statistics").get("numeric")[-2]
registered: dict[str, Any] = data_report.get("summary_statistics").get("numeric")[-1]
datetime: dict[str, Any] = data_report.get("summary_statistics").get("categorical")[0]

console.print(
    f"Numeric:\n{year}\n\n{atemp}\n\n{casual}\n\n{registered}\n\nCategorical: \n{datetime}"
)

Numeric:
{'column': 'yr', 'mean': 0.38, 'median': 0.0, 'mode': [0], 'std': 0.48, 'variance': 0.24, 'range': 1, 'min': 0, 
'max': 1, 'count': 13903, 'missing_values': 0, 'missing_pct': 0.0, 'unique_values': 2}

{'column': 'atemp', 'mean': 0.48, 'median': 0.48, 'mode': [0.6212], 'std': 0.18, 'variance': 0.03, 'range': 1.0, 
'min': 0.0, 'max': 1.0, 'count': 13903, 'missing_values': 0, 'missing_pct': 0.0, 'unique_values': 65}

{'column': 'registered', 'mean': 140.6, 'median': 107.0, 'mode': [4], 'std': 137.51, 'variance': 18909.65, 'range':
796, 'min': 0, 'max': 796, 'count': 13903, 'missing_values': 0, 'missing_pct': 0.0, 'unique_values': 701}

{'column': 'cnt', 'mean': 174.64, 'median': 130.0, 'mode': [5], 'std': 166.96, 'variance': 27874.11, 'range': 956, 
'min': 1, 'max': 957, 'count': 13903, 'missing_values': 0, 'missing_pct': 0.0, 'unique_values': 789}

Categorical: 
{'column': 'datetime', 'total_count': 13903, 'unique_values': 13903, 'value_counts': [['2011-01-01 00:00:00', 1], 
['2011-01-01 01:00:00', 1], ['2011-01-01 02:00:00', 1], ['2011-01-01 03:00:00', 1], ['2011-01-01 04:00:00', 1], 
['2011-01-01 05:00:00', 1], ['2011-01-01 06:00:00', 1], ['2011-01-01 07:00:00', 1], ['2011-01-01 08:00:00', 1], 
['2011-01-01 09:00:00', 1]], 'missing_values': 0, 'missing_pct': 0.0}

In [9]:
columns_to_drop: list[str] = ["yr", "atemp", "casual", "registered", "datetime"]
df: pl.DataFrame = data.clone().drop(columns_to_drop)

df.head(3)

shape: (3, 11)
┌────────┬──────┬─────┬─────────┬─────────┬────────────┬───────────┬──────┬──────┬───────────┬─────┐
│ season ┆ mnth ┆ hr  ┆ holiday ┆ weekday ┆ workingday ┆ weathersi ┆ temp ┆ hum  ┆ windspeed ┆ cnt │
│ ---    ┆ ---  ┆ --- ┆ ---     ┆ ---     ┆ ---        ┆ t         ┆ ---  ┆ ---  ┆ ---       ┆ --- │
│ i64    ┆ i64  ┆ i64 ┆ i64     ┆ i64     ┆ i64        ┆ ---       ┆ f64  ┆ f64  ┆ f64       ┆ i64 │
│        ┆      ┆     ┆         ┆         ┆            ┆ i64       ┆      ┆      ┆           ┆     │
╞════════╪══════╪═════╪═════════╪═════════╪════════════╪═══════════╪══════╪══════╪═══════════╪═════╡
│ 1      ┆ 1    ┆ 0   ┆ 0       ┆ 6       ┆ 0          ┆ 1         ┆ 0.24 ┆ 0.81 ┆ 0.0       ┆ 16  │
│ 1      ┆ 1    ┆ 1   ┆ 0       ┆ 6       ┆ 0          ┆ 1         ┆ 0.22 ┆ 0.8  ┆ 0.0       ┆ 40  │
│ 1      ┆ 1    ┆ 2   ┆ 0       ┆ 6       ┆ 0          ┆ 1         ┆ 0.22 ┆ 0.8  ┆ 0.0       ┆ 32  │
└────────┴──────┴─────┴─────────┴─────────┴────────────┴───────────┴──────┴──────┴───────────┴─────┘

In [10]:
# Since we want to maintain the temporal order, we'll
# use a custom function for splitting the data

# it's assumed the data has been sorted in ascending order of time
# get the train, val, test ratios
# -10% of the training data is used for validation
train_size: float = 0.9
data_array: np.ndarray = df.to_numpy()
train_array: np.ndarray = data_array[: int(train_size * data_array.shape[0])]
test_array: np.ndarray = data_array[train_array.shape[0] :]

train_array.shape, test_array.shape
# convert the ratios to int values and select a slice of the data corresponding to the ratio

((12512, 11), (1391, 11))

In [ ]:
from sklearn.metrics._regression import mean_absolute_error, root_mean_squared_error


def split_temporal_data(
    data: pl.DataFrame, test_size: float = 0.2
) -> tuple[pl.DataFrame, pl.DataFrame]:
    feature_names: list[str] = data.columns
    data_array: np.ndarray = data.to_numpy()
    train_size: float = int((1 - test_size) * data_array.shape[0])

    train_array: np.ndarray = data_array[:train_size]
    test_array: np.ndarray = data_array[train_size:]

    return pl.DataFrame(train_array, schema=feature_names), pl.DataFrame(
        test_array, schema=feature_names
    )


def compute_metrics(
    y_true: np.ndarray | list, y_pred: np.ndarray | list
) -> dict[str, float]:
    """
    Compute evaluation metrics between true and predicted values.

    Metrics returned:
    - MAPE: Mean Absolute Percentage Error (in %)
    - MAE: Mean Absolute Error
    - RMSE: Root Mean Squared Error

    Parameters:
    ----------
    y_true : array-like
        Ground truth values.
    y_pred : array-like
        Predicted values.

    Returns:
    -------
    dict
        Dictionary with keys 'MAPE', 'MAE', and 'RMSE' and their float values.
    """
    mae = mean_absolute_error(y_true, y_pred)
    rmse = root_mean_squared_error(y_true, y_pred)

    y_true = np.array(y_true)
    y_pred = np.array(y_pred)

    mape: float = (
        np.mean(np.abs((y_true - y_pred) / np.where(y_true == 0, 0.01, y_true))) * 100
    ).item()

    return {
        "MAE": round(mae, 2),
        "RMSE": round(rmse, 2),
        "MAPE": round(mape, 2),
    }

### Target Definition

- We'll use the next hour `cnt` as the target variable to predict bike rentals.
- This is because bike rental services typically operate on an hourly basis, and predicting the next hour's demand can help in resource allocation and planning.
- We'll shift the `cnt` column by one hour to create the target variable, ensuring that our model learns to predict future demand based on current and past data.
- i.e. `df["target"] = df["cnt"].shift(-1)` if `cnt`=[5, 10, 15, 20] then `target`=[10, 15, 20, NaN]
  - i.e. instead of predicting the current hour's demand, we are training the model to predict the demand for the next hour based on the current and past data.

In [12]:
df = df.with_columns(pl.col("cnt").shift(-1).alias("target")).with_columns(
    pl.col("target").fill_null(strategy="forward")
)

df.head(3)

shape: (3, 12)
┌────────┬──────┬─────┬─────────┬─────────┬─────────┬────────┬──────┬──────┬────────┬─────┬────────┐
│ season ┆ mnth ┆ hr  ┆ holiday ┆ weekday ┆ working ┆ weathe ┆ temp ┆ hum  ┆ windsp ┆ cnt ┆ target │
│ ---    ┆ ---  ┆ --- ┆ ---     ┆ ---     ┆ day     ┆ rsit   ┆ ---  ┆ ---  ┆ eed    ┆ --- ┆ ---    │
│ i64    ┆ i64  ┆ i64 ┆ i64     ┆ i64     ┆ ---     ┆ ---    ┆ f64  ┆ f64  ┆ ---    ┆ i64 ┆ i64    │
│        ┆      ┆     ┆         ┆         ┆ i64     ┆ i64    ┆      ┆      ┆ f64    ┆     ┆        │
╞════════╪══════╪═════╪═════════╪═════════╪═════════╪════════╪══════╪══════╪════════╪═════╪════════╡
│ 1      ┆ 1    ┆ 0   ┆ 0       ┆ 6       ┆ 0       ┆ 1      ┆ 0.24 ┆ 0.81 ┆ 0.0    ┆ 16  ┆ 40     │
│ 1      ┆ 1    ┆ 1   ┆ 0       ┆ 6       ┆ 0       ┆ 1      ┆ 0.22 ┆ 0.8  ┆ 0.0    ┆ 40  ┆ 32     │
│ 1      ┆ 1    ┆ 2   ┆ 0       ┆ 6       ┆ 0       ┆ 1      ┆ 0.22 ┆ 0.8  ┆ 0.0    ┆ 32  ┆ 13     │
└────────┴──────┴─────┴─────────┴─────────┴─────────┴────────┴──────┴──────┴────────┴─────┴────────┘

In [13]:
df.null_count()

shape: (1, 12)
┌────────┬──────┬─────┬─────────┬─────────┬─────────┬─────────┬──────┬─────┬────────┬─────┬────────┐
│ season ┆ mnth ┆ hr  ┆ holiday ┆ weekday ┆ working ┆ weather ┆ temp ┆ hum ┆ windsp ┆ cnt ┆ target │
│ ---    ┆ ---  ┆ --- ┆ ---     ┆ ---     ┆ day     ┆ sit     ┆ ---  ┆ --- ┆ eed    ┆ --- ┆ ---    │
│ u32    ┆ u32  ┆ u32 ┆ u32     ┆ u32     ┆ ---     ┆ ---     ┆ u32  ┆ u32 ┆ ---    ┆ u32 ┆ u32    │
│        ┆      ┆     ┆         ┆         ┆ u32     ┆ u32     ┆      ┆     ┆ u32    ┆     ┆        │
╞════════╪══════╪═════╪═════════╪═════════╪═════════╪═════════╪══════╪═════╪════════╪═════╪════════╡
│ 0      ┆ 0    ┆ 0   ┆ 0       ┆ 0       ┆ 0       ┆ 0       ┆ 0    ┆ 0   ┆ 0      ┆ 0   ┆ 0      │
└────────┴──────┴─────┴─────────┴─────────┴─────────┴─────────┴──────┴─────┴────────┴─────┴────────┘

In [14]:
df = df.drop("cnt")

train_df, test_df = split_temporal_data(df)

x_train = train_df.drop("target").to_numpy()
y_train = train_df["target"].to_numpy()

x_test = test_df.drop("target").to_numpy()
y_test = test_df["target"].to_numpy()

x_train.shape, y_train.shape, x_test.shape, y_test.shape

((11122, 10), (11122,), (2781, 10), (2781,))

<br>

# Baseline Model

- Create a simple and basic ML model to be used as reference.
- I used a `RandomForestRegressor` because it requires little tuning.

In [15]:
from sklearn.ensemble import RandomForestRegressor

rf_reg = RandomForestRegressor(random_state=123)
rf_reg.fit(x_train, y_train)
rf_reg.score(x_test, y_test)

y_pred = rf_reg.predict(x_test)

In [20]:
metrics_base = compute_metrics(y_test, y_pred)
metrics_base

{'MAE': 84.78, 'RMSE': 116.41, 'MAPE': 34.11}

### Comment

- The model tends to underpredict during most of the high-peak periods.
- This suggests that the model is underestimating the number of bike rentals.

In [22]:
import plotly.graph_objects as go

n: int = 300

fig = go.Figure()

# Add actual values trace
fig.add_trace(
    go.Scatter(
        x=list(range(n)),
        y=y_test[:n],
        mode="lines",
        name="y_test (Actual)",
        line={"color": "blue", "width": 2},
    )
)

# Add predicted values trace
fig.add_trace(
    go.Scatter(
        x=list(range(n)),
        y=y_pred[:n],
        mode="lines",
        name="y_pred (Predicted)",
        line={"color": "red", "width": 2},
    )
)

# Update layout for better visualization
fig.update_layout(
    title="Actual vs Predicted Bike Rental Counts",
    xaxis_title="Time Steps",
    yaxis_title="Bike Rental Count",
    hovermode="x unified",
    template="plotly_white",
    width=800,
    height=500,
)

# Show the interactive plot
fig.show()

### Dummy Model

- Let's create a dummy model that predicts the next hour's bike rentals based on exactly the same hour of the previous day.

In [26]:
y_dummy = pl.Series(y_test).shift(1).fill_null(strategy="backward").to_numpy()
y_dummy

array([463., 463., 738., ..., 327., 250., 214.], shape=(2781,))

In [27]:
metrics_dummy: dict[str, float] = compute_metrics(y_test, y_dummy)

console.print(f"Metrics (Base Model): {metrics_base}", style="info")
console.print(f"Metrics (Dummy Model): {metrics_dummy}", style="info")

Metrics (Base Model): {'MAE': 84.78, 'RMSE': 116.41, 'MAPE': 34.11}

Metrics (Dummy Model): {'MAE': 87.01, 'RMSE': 128.17, 'MAPE': 55.4}

### Comment

- The baseline model's performance is not a significant improvement over the dummy model. 

- While the Mean Absolute Percentage Error (`MAPE`) improved notably, the Mean Absolute Error (`MAE`) and other metrics show little change, suggesting that feature engineering is necessary to improve the model.

In [36]:
def pct_change(old: float, new: float) -> float:
    """Calculate the percentage change from old to new value."""
    if old == 0:
        return float("inf")  # Avoid division by zero
    return (abs(new - old) / old) * 100


mae_base, mae_dummy = metrics_base.get("MAE"), metrics_dummy.get("MAE")
rmse_base, rmse_dummy = metrics_base.get("RMSE"), metrics_dummy.get("RMSE")
mape_base, mape_dummy = metrics_base.get("MAPE"), metrics_dummy.get("MAPE")

console.print(
    f"MAE Change: {pct_change(old=mae_base, new=mae_dummy):.2f}%", style="highlight"
)
console.print(
    f"RMSE Change: {pct_change(old=rmse_base, new=rmse_dummy):.2f}%", style="highlight"
)
console.print(
    f"MAPE Change: {pct_change(old=mape_base, new=mape_dummy):.2f}%", style="highlight"
)

MAE Change: 2.63%

RMSE Change: 10.10%

MAPE Change: 62.42%

# Baseline Model With Lagged Target Features

- Lagged features are past values of a time series used to capture autocorrelation (i.e. the relationship between a data point and its previous values). 

- This is based on the idea that the past influences the future.

- By including lagged features, the model gains a historical context that helps it identify:

  - `Trends`: Long-term upward or downward movements.

  - `Seasonality`: Regular, repeating patterns, such as daily or weekly cycles.

  - `Temporal dependencies`: How a value at one time step is directly related to its value at a previous time step.

In [70]:
import narwhals as nw


def _calculate_corr(x: list[float] | np.ndarray, y: list[float] | np.ndarray) -> float:
    return np.corrcoef(x, y)[0][1].item()


def compute_autocorrelation(series: nw.Series, max_lag: int = 24) -> dict[int, float]:
    """
    Compute autocorrelation for a time series using Polars.

    Parameters
    ----------
    series : pl.Series
        The time series data.
    max_lag : int, default=24
        Maximum lag to compute autocorrelation for.

    Returns
    -------
    dict[int, float]
        Dictionary mapping lag to autocorrelation value.
    """
    autocorr_values: dict[int, float] = {}

    for lag in range(1, max_lag + 1):
        try:
            # Create a DataFrame with original and lagged series
            df_corr: nw.DataFrame = nw.from_native(
                pl.DataFrame({"original": series, "lagged": series.shift(lag)})
            ).drop_nulls()

            # Compute correlation if we have sufficient data
            if df_corr.shape[0] > 1:
                correlation = _calculate_corr(df_corr["original"], df_corr["lagged"])
                autocorr_values[lag] = correlation
            else:
                autocorr_values[lag] = None

        except Exception as e:
            console.print(f"Error computing lag {lag}: {e}", style="error")
            autocorr_values[lag] = None

    return autocorr_values

In [121]:
# Compute autocorrelation of the target variable
auto_correlation: dict[int, float] = compute_autocorrelation(
    series=df["target"], max_lag=24
)
for lag, autocorr in auto_correlation.items():
    console.print(f"Lag {lag}: Autocorrelation = {autocorr:.3f}", style="info")

Lag 1: Autocorrelation = 0.845

Lag 2: Autocorrelation = 0.597

Lag 3: Autocorrelation = 0.407

Lag 4: Autocorrelation = 0.274

Lag 5: Autocorrelation = 0.146

Lag 6: Autocorrelation = 0.010

Lag 7: Autocorrelation = -0.094

Lag 8: Autocorrelation = -0.114

Lag 9: Autocorrelation = -0.058

Lag 10: Autocorrelation = -0.049

Lag 11: Autocorrelation = -0.100

Lag 12: Autocorrelation = -0.131

Lag 13: Autocorrelation = -0.105

Lag 14: Autocorrelation = -0.057

Lag 15: Autocorrelation = -0.058

Lag 16: Autocorrelation = -0.094

Lag 17: Autocorrelation = -0.065

Lag 18: Autocorrelation = 0.033

Lag 19: Autocorrelation = 0.153

Lag 20: Autocorrelation = 0.262

Lag 21: Autocorrelation = 0.371

Lag 22: Autocorrelation = 0.520

Lag 23: Autocorrelation = 0.705

Lag 24: Autocorrelation = 0.809

### Comment

- **Strong Hourly Dependency**: The high correlation at Lag 1 (r=0.845) shows that bike rentals are highly dependent on the previous hour's count.

- **Clear Daily Seasonality**: The strong correlation at Lag 24 (r=0.809) confirms a predictable daily pattern in bike usage.

- **Inverse Peak/Off-Peak Relationship**: The negative correlations between 6 and 18 hours apart indicate that peak and off-peak periods are inversely related.

- **Modeling Strategy**: Lagged features (especially 1, 2, 23, and 24 hours) and time-of-day features will be essential for building an effective model.

In [110]:
columns_to_drop: list[str] = ["yr", "atemp", "casual", "registered", "datetime"]
df: pl.DataFrame = data.clone().drop(columns_to_drop)

df.head(3)

shape: (3, 11)
┌────────┬──────┬─────┬─────────┬─────────┬────────────┬───────────┬──────┬──────┬───────────┬─────┐
│ season ┆ mnth ┆ hr  ┆ holiday ┆ weekday ┆ workingday ┆ weathersi ┆ temp ┆ hum  ┆ windspeed ┆ cnt │
│ ---    ┆ ---  ┆ --- ┆ ---     ┆ ---     ┆ ---        ┆ t         ┆ ---  ┆ ---  ┆ ---       ┆ --- │
│ i64    ┆ i64  ┆ i64 ┆ i64     ┆ i64     ┆ i64        ┆ ---       ┆ f64  ┆ f64  ┆ f64       ┆ i64 │
│        ┆      ┆     ┆         ┆         ┆            ┆ i64       ┆      ┆      ┆           ┆     │
╞════════╪══════╪═════╪═════════╪═════════╪════════════╪═══════════╪══════╪══════╪═══════════╪═════╡
│ 1      ┆ 1    ┆ 0   ┆ 0       ┆ 6       ┆ 0          ┆ 1         ┆ 0.24 ┆ 0.81 ┆ 0.0       ┆ 16  │
│ 1      ┆ 1    ┆ 1   ┆ 0       ┆ 6       ┆ 0          ┆ 1         ┆ 0.22 ┆ 0.8  ┆ 0.0       ┆ 40  │
│ 1      ┆ 1    ┆ 2   ┆ 0       ┆ 6       ┆ 0          ┆ 1         ┆ 0.22 ┆ 0.8  ┆ 0.0       ┆ 32  │
└────────┴──────┴─────┴─────────┴─────────┴────────────┴───────────┴──────┴──────┴───────────┴─────┘

In [111]:
# Shift the `cnt` column by one hour to create the target variable
# i.e. predict the next hour's demand based on current and past data
df = df.with_columns(pl.col("cnt").shift(-1).alias("target")).with_columns(
    pl.col("target").fill_null(strategy="forward")
)
df.head()

shape: (5, 12)
┌────────┬──────┬─────┬─────────┬─────────┬─────────┬────────┬──────┬──────┬────────┬─────┬────────┐
│ season ┆ mnth ┆ hr  ┆ holiday ┆ weekday ┆ working ┆ weathe ┆ temp ┆ hum  ┆ windsp ┆ cnt ┆ target │
│ ---    ┆ ---  ┆ --- ┆ ---     ┆ ---     ┆ day     ┆ rsit   ┆ ---  ┆ ---  ┆ eed    ┆ --- ┆ ---    │
│ i64    ┆ i64  ┆ i64 ┆ i64     ┆ i64     ┆ ---     ┆ ---    ┆ f64  ┆ f64  ┆ ---    ┆ i64 ┆ i64    │
│        ┆      ┆     ┆         ┆         ┆ i64     ┆ i64    ┆      ┆      ┆ f64    ┆     ┆        │
╞════════╪══════╪═════╪═════════╪═════════╪═════════╪════════╪══════╪══════╪════════╪═════╪════════╡
│ 1      ┆ 1    ┆ 0   ┆ 0       ┆ 6       ┆ 0       ┆ 1      ┆ 0.24 ┆ 0.81 ┆ 0.0    ┆ 16  ┆ 40     │
│ 1      ┆ 1    ┆ 1   ┆ 0       ┆ 6       ┆ 0       ┆ 1      ┆ 0.22 ┆ 0.8  ┆ 0.0    ┆ 40  ┆ 32     │
│ 1      ┆ 1    ┆ 2   ┆ 0       ┆ 6       ┆ 0       ┆ 1      ┆ 0.22 ┆ 0.8  ┆ 0.0    ┆ 32  ┆ 13     │
│ 1      ┆ 1    ┆ 3   ┆ 0       ┆ 6       ┆ 0       ┆ 1      ┆ 0.24 ┆ 0.75 ┆ 0.0    ┆ 13  ┆ 1      │
│ 1      ┆ 1    ┆ 4   ┆ 0       ┆ 6       ┆ 0       ┆ 1      ┆ 0.24 ┆ 0.75 ┆ 0.0    ┆ 1   ┆ 1      │
└────────┴──────┴─────┴─────────┴─────────┴─────────┴────────┴──────┴──────┴────────┴─────┴────────┘

In [112]:
# Add lags 1, 23 and 24 hours
df = df.with_columns(
    [
        pl.col("cnt").alias("current"),
        pl.col("cnt").shift(1).alias("lag_1"),
        pl.col("cnt").shift(23).alias("lag_23"),
        pl.col("cnt").shift(24).alias("lag_24"),
    ]
).fill_null(strategy="backward")
print("Check for null values after adding lag features:")
display(df.null_count())
df.head()

Check for null values after adding lag features:


shape: (1, 16)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┐
│ sea ┆ mnt ┆ hr  ┆ hol ┆ wee ┆ wor ┆ wea ┆ tem ┆ hum ┆ win ┆ cnt ┆ tar ┆ cur ┆ lag ┆ lag ┆ lag_24 │
│ son ┆ h   ┆ --- ┆ ida ┆ kda ┆ kin ┆ the ┆ p   ┆ --- ┆ dsp ┆ --- ┆ get ┆ ren ┆ _1  ┆ _23 ┆ ---    │
│ --- ┆ --- ┆ u32 ┆ y   ┆ y   ┆ gda ┆ rsi ┆ --- ┆ u32 ┆ eed ┆ u32 ┆ --- ┆ t   ┆ --- ┆ --- ┆ u32    │
│ u32 ┆ u32 ┆     ┆ --- ┆ --- ┆ y   ┆ t   ┆ u32 ┆     ┆ --- ┆     ┆ u32 ┆ --- ┆ u32 ┆ u32 ┆        │
│     ┆     ┆     ┆ u32 ┆ u32 ┆ --- ┆ --- ┆     ┆     ┆ u32 ┆     ┆     ┆ u32 ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆ u32 ┆ u32 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╡
│ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0      │
└─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴────────┘

shape: (5, 16)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬────────┐
│ sea ┆ mnt ┆ hr  ┆ hol ┆ wee ┆ wor ┆ wea ┆ tem ┆ hum ┆ win ┆ cnt ┆ tar ┆ cur ┆ lag ┆ lag ┆ lag_24 │
│ son ┆ h   ┆ --- ┆ ida ┆ kda ┆ kin ┆ the ┆ p   ┆ --- ┆ dsp ┆ --- ┆ get ┆ ren ┆ _1  ┆ _23 ┆ ---    │
│ --- ┆ --- ┆ i64 ┆ y   ┆ y   ┆ gda ┆ rsi ┆ --- ┆ f64 ┆ eed ┆ i64 ┆ --- ┆ t   ┆ --- ┆ --- ┆ i64    │
│ i64 ┆ i64 ┆     ┆ --- ┆ --- ┆ y   ┆ t   ┆ f64 ┆     ┆ --- ┆     ┆ i64 ┆ --- ┆ i64 ┆ i64 ┆        │
│     ┆     ┆     ┆ i64 ┆ i64 ┆ --- ┆ --- ┆     ┆     ┆ f64 ┆     ┆     ┆ i64 ┆     ┆     ┆        │
│     ┆     ┆     ┆     ┆     ┆ i64 ┆ i64 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪════════╡
│ 1   ┆ 1   ┆ 0   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 16  ┆ 40  ┆ 16  ┆ 16  ┆ 16  ┆ 16     │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 1   ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ 1   ┆ 1   ┆ 1   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 40  ┆ 32  ┆ 40  ┆ 16  ┆ 16  ┆ 16     │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 2   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ 1   ┆ 1   ┆ 2   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.8 ┆ 0.0 ┆ 32  ┆ 13  ┆ 32  ┆ 40  ┆ 16  ┆ 16     │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 2   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ 1   ┆ 1   ┆ 3   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.7 ┆ 0.0 ┆ 13  ┆ 1   ┆ 13  ┆ 32  ┆ 16  ┆ 16     │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 5   ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
│ 1   ┆ 1   ┆ 4   ┆ 0   ┆ 6   ┆ 0   ┆ 1   ┆ 0.2 ┆ 0.7 ┆ 0.0 ┆ 1   ┆ 1   ┆ 1   ┆ 13  ┆ 16  ┆ 16     │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 4   ┆ 5   ┆     ┆     ┆     ┆     ┆     ┆     ┆        │
└─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴────────┘

In [113]:
df = df.drop(["cnt"])

train_df, test_df = split_temporal_data(df)

x_train = train_df.drop("target").to_numpy()
y_train = train_df["target"].to_numpy()

x_test = test_df.drop("target").to_numpy()
y_test = test_df["target"].to_numpy()

x_train.shape, y_train.shape, x_test.shape, y_test.shape

((11122, 14), (11122,), (2781, 14), (2781,))

In [114]:
rf_reg = RandomForestRegressor(random_state=123)
rf_reg.fit(x_train, y_train)
rf_reg.score(x_test, y_test)

y_pred = rf_reg.predict(x_test)

In [115]:
metrics_base_with_lagged_feats = compute_metrics(y_test, y_pred)
metrics_base_with_lagged_feats

{'MAE': 43.39, 'RMSE': 70.61, 'MAPE': 24.28}

In [116]:
n: int = 300

fig = go.Figure()

# Add actual values trace
fig.add_trace(
    go.Scatter(
        x=list(range(n)),
        y=y_test[:n],
        mode="lines",
        name="y_test (Actual)",
        line={"color": "blue", "width": 2},
    )
)

# Add predicted values trace
fig.add_trace(
    go.Scatter(
        x=list(range(n)),
        y=y_pred[:n],
        mode="lines",
        name="y_pred (Predicted with Lagged Features)",
        line={"color": "red", "width": 2},
    )
)

# Update layout for better visualization
fig.update_layout(
    title="Actual vs Predicted Bike Rental Counts",
    xaxis_title="Time Steps",
    yaxis_title="Bike Rental Count",
    hovermode="x unified",
    template="plotly_white",
    width=800,
    height=500,
)

# Show the interactive plot
fig.show()

In [117]:
console.print(f"Metrics (Base Model): {metrics_base}", style="info")
console.print(f"Metrics (Dummy Model): {metrics_dummy}", style="info")
console.print(
    f"Metrics (Base Model with Lagged Features): {metrics_base_with_lagged_feats}",
    style="info",
)

Metrics (Base Model): {'MAE': 84.78, 'RMSE': 116.41, 'MAPE': 34.11}

Metrics (Dummy Model): {'MAE': 87.01, 'RMSE': 128.17, 'MAPE': 55.4}

Metrics (Base Model with Lagged Features): {'MAE': 43.39, 'RMSE': 70.61, 'MAPE': 24.28}

In [118]:
mae_base, mae_dummy = (
    metrics_base_with_lagged_feats.get("MAE"),
    metrics_dummy.get("MAE"),
)
rmse_base, rmse_dummy = (
    metrics_base_with_lagged_feats.get("RMSE"),
    metrics_dummy.get("RMSE"),
)
mape_base, mape_dummy = (
    metrics_base_with_lagged_feats.get("MAPE"),
    metrics_dummy.get("MAPE"),
)

console.print(
    f"MAE Change: {pct_change(old=mae_base, new=mae_dummy):.2f}%", style="highlight"
)
console.print(
    f"RMSE Change: {pct_change(old=rmse_base, new=rmse_dummy):.2f}%", style="highlight"
)
console.print(
    f"MAPE Change: {pct_change(old=mape_base, new=mape_dummy):.2f}%", style="highlight"
)

MAE Change: 100.53%

RMSE Change: 81.52%

MAPE Change: 128.17%